### PR TITLE
AML-2804 AML-2970 EmployerFeatures & ProviderPermissions authorization handlers

### PR DESCRIPTION
**Pull request by lazcool**

### CODE REVIEW

File: src/SFA.DAS.Authorization.ProviderPermissions.UnitTests/AuthorizationHandlerTests.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+using SFA.DAS.Testing;
+
+namespace SFA.DAS.Authorization.ProviderPermissions.UnitTests
+{
+    [TestFixture]
+    [Parallelizable]
+    public class AuthorizationHandlerTests : FluentTest<AuthorizationHandlerTestsFixture>
+    {
+        [Test]
+        public Task GetAuthorizationResultAsync_WhenOptionsAreNotAvailable_ThenShouldReturnValidAuthorizationResult()
```
Comment (Pauljgraham) [inline src/SFA.DAS.Authorization.ProviderPermissions.UnitTests/AuthorizationHandlerTests.cs]: Simple rename to GetAuthorizationResult dropping the Async word as the function is now called this

File: src/SFA.DAS.Authorization.EmployerFeatures/FeatureToggle.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            Feature = feature;
+            IsEnabled = isEnabled;
+            UserEmailWhitelist = userEmailWhitelist;
+        }
+
+        public bool IsUserEmailWhitelisted(string userEmail)
+        {
+            return UserEmailWhitelist.Any() && UserEmailWhitelist.Any(e => e.Equals(userEmail, StringComparison.InvariantCultureIgnoreCase));
+        }
```
Comment (Pauljgraham) [inline src/SFA.DAS.Authorization.EmployerFeatures/FeatureToggle.cs]: Why have UserEmailWhitelist.Any() preceeded the UserEmailWhitelist.Any(....

File: src/SFA.DAS.Authorization.EmployerFeatures/AuthorizationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+                var values = authorizationContext.GetEmployerFeatureValues();
+                var feature = options.Select(o => o.ToEnum<Feature>()).Single();
+                var featureToggle = await _featureTogglesService.GetFeatureToggle(feature).ConfigureAwait(false);
+
+                if (!featureToggle.IsEnabled)
+                {
+                    authorizationResult.AddError(new EmployerFeatureNotEnabled());
+                }
+                else if (!featureToggle.IsUserEmailWhitelisted(values.UserEmail))
```
Comment (Pauljgraham) [inline src/SFA.DAS.Authorization.EmployerFeatures/AuthorizationHandler.cs]: If the feature is enabled and there is no white listing, will the user see the feature. Looking at the code it looks like if there is no 'white listing' the feature will not be visible

File: src/SFA.DAS.Authorization.EmployerFeatures.UnitTests/FeatureTogglesServiceTests.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+namespace SFA.DAS.Authorization.EmployerFeatures.UnitTests
+{
+    [TestFixture]
+    public class FeatureTogglesServiceTests : FluentTest<FeatureTogglesServiceTestsFixture>
+    {
+        [Test]
+        public Task GetFeatureToggle_WhenGettingFeatureToggle_ThenShouldReturnFeatureToggle()
+        {
+            return RunAsync(f => f.GetFeatureToggle(), (f, r) => r.Should().NotBeNull().And.BeSameAs(f.FeatureToggle));
+        }
```
[the remark this answers is not in the document]
Comment (elliotritchie) [reply src/SFA.DAS.Authorization.EmployerFeatures.UnitTests/FeatureTogglesServiceTests.cs]: Yeah, it should return a disabled feature toggle. I've added a unit test to cover the scenario.

File: src/SFA.DAS.Authorization.EmployerFeatures/FeatureTogglesService.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+        public FeatureTogglesService(EmployerFeaturesConfiguration configuration)
+        {
+            _featureToggles = Enum.GetValues(typeof(Feature)).Cast<Feature>().ToDictionary(f => f, f => 
+                configuration.FeatureToggles.SingleOrDefault(t => t.Feature == f) ?? new FeatureToggle(f, false, new List<string>()));
+        }
+
+        public Task<FeatureToggle> GetFeatureToggle(Feature feature)
+        {
+            return Task.FromResult(_featureToggles[feature]);
```
[the remark this answers is not in the document]
Comment (elliotritchie) [reply src/SFA.DAS.Authorization.EmployerFeatures/FeatureTogglesService.cs]: Fixed

File: src/SFA.DAS.Authorization.ProviderPermissions.UnitTests/AuthorizationHandlerTests.cs (hunk 1 above)
Comment (elliotritchie) [reply src/SFA.DAS.Authorization.ProviderPermissions.UnitTests/AuthorizationHandlerTests.cs]: Fixed

File: src/SFA.DAS.Authorization.EmployerFeatures/FeatureToggle.cs (hunk 2 above)
Comment (elliotritchie) [reply src/SFA.DAS.Authorization.EmployerFeatures/FeatureToggle.cs]: Fixed

File: src/SFA.DAS.Authorization.EmployerFeatures/AuthorizationHandler.cs (hunk 3 above)
Comment (elliotritchie) [reply src/SFA.DAS.Authorization.EmployerFeatures/AuthorizationHandler.cs]: Yeah, the user should see the feature if the toggle is enabled and the user email whitelist is empty.

File: src/SFA.DAS.Authorization.EmployerRoles.UnitTests/AuthorizationHandlerTests.cs [hunk 6]
```diff
@@ -24,7 +25,7 @@ public Task GetAuthorizationResultAsync_WhenGettingAuthorizationResultAndNonEmpl
         [Test]
         public Task GetAuthorizationResultAsync_WhenGettingAuthorizationResultAndEmployerRolesOptionsAreAvailableAndEmployerRoleContextIsNotAvailable_ThenShouldThrowAuthorizationContextKeyNotFoundException()
         {
-            return RunAsync(f => f.SetEmployerRolesOptions().SetNonEmployerRolesContext(), f => f.GetAuthorizationResultAsync(), (f, r) => r.Should().Throw<KeyNotFoundException>());
+            return RunAsync(f => f.SetEmployerRolesOptions(), f => f.GetAuthorizationResultAsync(), (f, r) => r.Should().Throw<KeyNotFoundException>());
```
[the remark this answers is not in the document]
Comment (elliotritchie) [reply src/SFA.DAS.Authorization.EmployerRoles.UnitTests/AuthorizationHandlerTests.cs]: Fixed

**Pull request merged**